Upgrade modules and support go 1.20

ToxicGLaDOS opened the pull request:
Closes #774
In [v0.32.0](https://github.com/quic-go/quic-go/releases/tag/v0.32.0) of quic-go they switched to their own organization instead of lucas-clemente's personal account, hence the change to pkg/netceptor/conn.go.

ToxicGLaDOS committed:
Upgrade modules and support go 1.20

Closes #774
In [v0.32.0](https://github.com/quic-go/quic-go/releases/tag/v0.32.0) of
quic-go they switched to their own organization instead of
lucas-clemente's personal account, hence the change to pkg/netceptor/conn.go.

ToxicGLaDOS commented:
quic-go dropped support for Go 1.18 in [v0.33.0](https://github.com/quic-go/quic-go/releases/tag/v0.33.0), so it might make sense to update the automation to check against Go 1.20 instead of Go 1.18 in this pr too.

AaronH88 commented:
Thanks for the PR,

Can you change this file to [1.19, 1.20] please? : https://github.com/ansible/receptor/blob/devel/.github/workflows/pull_request.yml#L42

ToxicGLaDOS committed:
Update pull request workflow to use golang 1.19 and 1.20

ToxicGLaDOS commented:
> Thanks for the PR,
> 
> Can you change this file to [1.19, 1.20] please? : https://github.com/ansible/receptor/blob/devel/.github/workflows/pull_request.yml#L42

Done! But do we also want to update [build_binary_from_ref.yml](https://github.com/ansible/receptor/blob/devel/.github/workflows/build_binary_from_ref.yml#LL21C5-L21C5)
[promote.yml](https://github.com/ansible/receptor/blob/devel/.github/workflows/promote.yml#L85) to 1.20? 1.19 should still work, but it seems we were doing the latest working version previously.

AaronH88 commented:
> > Thanks for the PR,
> > Can you change this file to [1.19, 1.20] please? : https://github.com/ansible/receptor/blob/devel/.github/workflows/pull_request.yml#L42
> 
> Done! But do we also want to update [build_binary_from_ref.yml](https://github.com/ansible/receptor/blob/devel/.github/workflows/build_binary_from_ref.yml#LL21C5-L21C5) [promote.yml](https://github.com/ansible/receptor/blob/devel/.github/workflows/promote.yml#L85) to 1.20? 1.19 should still work, but it seems we were doing the latest working version previously.

I think your correct, I missed that part!
Are you ok to add this?

ToxicGLaDOS commented:
@AaronH88 Yeah, I can add that. Do you know why the CI is saying "Go 1.2" instead of "Go 1.20"?

AaronH88 commented:
> @AaronH88 Yeah, I can add that. Do you know why the CI is saying "Go 1.2" instead of "Go 1.20"?

Thanks!

No, I am currently searching now to see if the `go-version` field is being treated as a float

ToxicGLaDOS committed:
Update promote and build binary from ref workflows to use go 1.20

AaronH88 commented on .github/workflows/pull_request.yml:
```diff
@@ -39,7 +39,7 @@ jobs:
       fail-fast: false
       matrix:
-        go-version: [1.18, 1.19]
+        go-version: [1.19, 1.20]
```
From https://docs.ansible.com/ansible/latest/reference_appendices/YAMLSyntax.html

YAML converts certain strings into floating-point values, such as the string 1.0. If you need to specify a version number (in a requirements.yml file, for example), you will need to quote the value if it looks like a floating-point value:

version: "1.0"

Can we put 1.19 and 1.20 in quotes please

ToxicGLaDOS committed:
Quote values to prevent yaml interpreting as a float

AaronH88 commented:
Looking great @ToxicGLaDOS , thanks so much!

The linter has shown some issues now that we are using 1.20 . This will explode the size of this PR.

If you dont have time to go through these linting issues, thats all good, I can take over from here later in the week.
Let me know how youd like to proceed, and thanks again! :smile:

ToxicGLaDOS commented:
@AaronH88 I should be able to fix the linter issues as long as they don't require any structural changes to the code. If I run out of time to work on this I'll post again and let ya know.

ToxicGLaDOS committed:
Fix new go 1.20 linter issues

ToxicGLaDOS commented:
@AaronH88 Okay! I pushed some updates, but there were a few changes that I'm unsure about that I'd like someone else who's more familiar with the codebase to look at. Specifically

1. The changes to `_` for unused variables, can those parameters just be removed from the function definition?
2. Is [Context.Background()](https://github.com/ansible/receptor/pull/779/commits/eebb57c87b629832bfcf2dc68f85ad787e6f015b#diff-792611baeb730234abface9c9bc33204ea86453469e7fcdfbe8de8d4e04f2598R455) the correct context?
3. I capitalized [BackendInfo](https://github.com/ansible/receptor/pull/779/commits/eebb57c87b629832bfcf2dc68f85ad787e6f015b#diff-34e0b90451affd6f30e2a36919a995c319236ef547e1d3e44c060d1700bd01deR269) to export it but we could've also not exported the functions that returned it (`TCPListen`, `UDPListen`, and `WebsocketListen`) as far as I can tell we don't use those functions outside of the package, but maybe it's exported for external use 🤷 
4. I only explicitly ignored [one linter issue](https://github.com/ansible/receptor/pull/779/commits/eebb57c87b629832bfcf2dc68f85ad787e6f015b#diff-e014010fd9a9e6fa761fb46bc1ca5f62bea8b3b699cec05c8ae75c3296b47834R186) because the comment made it sound like that's what we wanted. I'd just like to confirm that that's still true.

AaronH88 commented:
> @AaronH88 Okay! I pushed some updates, but there were a few changes that I'm unsure about that I'd like someone else who's more familiar with the codebase to look at. Specifically
> 
> 1. The changes to `_` for unused variables, can those parameters just be removed from the function definition?
> 2. Is [Context.Background()](https://github.com/ansible/receptor/pull/779/commits/eebb57c87b629832bfcf2dc68f85ad787e6f015b#diff-792611baeb730234abface9c9bc33204ea86453469e7fcdfbe8de8d4e04f2598R455) the correct context?
> 3. I capitalized [BackendInfo](https://github.com/ansible/receptor/pull/779/commits/eebb57c87b629832bfcf2dc68f85ad787e6f015b#diff-34e0b90451affd6f30e2a36919a995c319236ef547e1d3e44c060d1700bd01deR269) to export it but we could've also not exported the functions that returned it (`TCPListen`, `UDPListen`, and `WebsocketListen`) as far as I can tell we don't use those functions outside of the package, but maybe it's exported for external use shrug
> 4. I only explicitly ignored [one linter issue](https://github.com/ansible/receptor/pull/779/commits/eebb57c87b629832bfcf2dc68f85ad787e6f015b#diff-e014010fd9a9e6fa761fb46bc1ca5f62bea8b3b699cec05c8ae75c3296b47834R186) because the comment made it sound like that's what we wanted. I'd just like to confirm that that's still true.

Amazing!

Ill reply in line :smile:

AaronH88 commented on cmd/receptor-cl/receptor.go:
```diff
@@ -20,7 +20,7 @@ import (
 
 // make the nullBackendCfg object be usable as a do-nothing Backend.
-func (cfg nullBackendCfg) Start(ctx context.Context, wg *sync.WaitGroup) (chan netceptor.BackendSession, error) {
+func (cfg nullBackendCfg) Start(_ context.Context, _ *sync.WaitGroup) (chan netceptor.BackendSession, error) {
```
`Start` is an interface so this is correct!

AaronH88 commented on pkg/controlsvc/connect.go:
```diff
@@ -74,7 +74,7 @@ func (t *connectCommandType) InitFromJSON(config map[string]interface{}) (Contro
 }
 
-func (c *connectCommand) ControlFunc(ctx context.Context, nc *netceptor.Netceptor, cfo ControlFuncOperations) (map[string]interface{}, error) {
+func (c *connectCommand) ControlFunc(_ context.Context, nc *netceptor.Netceptor, cfo ControlFuncOperations) (map[string]interface{}, error) {
```
All `ControlFunc` are interfaces so leaving the signature alone is perfect.

This however highlights that we dont check the context for `ctx.Done()`.
I will need to add this to the backlog as it is outside the scope of this PR

AaronH88 commented on pkg/controlsvc/reload.go:
```diff
@@ -137,13 +137,13 @@ func checkReload() error {
 }
 
-func (t *reloadCommandType) InitFromString(params string) (ControlCommand, error) {
+func (t *reloadCommandType) InitFromString(_ string) (ControlCommand, error) {
```
both `Init*` 's are interfaces so this is also correct

AaronH88 commented on pkg/netceptor/conn.go:
```diff
@@ -490,7 +490,7 @@ func generateServerTLSConfig() *tls.Config {
 }
 
-func verifyServerCertificate(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
+func verifyServerCertificate(rawCerts [][]byte, _ [][]*x509.Certificate) error {
```
this is also an implementation of an interface, the `tls.Config` interface in crypto.go, so again, this is correct

AaronH88 commented on pkg/netceptor/external_backend.go:
```diff
@@ -92,7 +92,7 @@ func (mc *netMessageConn) ReadMessage(ctx context.Context, timeout time.Duration
 
 // SetReadDeadline sets the deadline by which a message must be read from the connection.
-func (mc *netMessageConn) SetReadDeadline(t time.Time) error {
+func (mc *netMessageConn) SetReadDeadline(_ time.Time) error {
```
Interface, this is correct

AaronH88 commented on pkg/netceptor/external_backend.go:
```diff
@@ -123,7 +123,7 @@ func (mc *websocketMessageConn) WriteMessage(ctx context.Context, data []byte) e
 
 // ReadMessage reads a message from the connection.
-func (mc *websocketMessageConn) ReadMessage(ctx context.Context, timeout time.Duration) ([]byte, error) {
+func (mc *websocketMessageConn) ReadMessage(ctx context.Context, _ time.Duration) ([]byte, error) {
```
Interface, this is correct

AaronH88 commented on pkg/netceptor/packetconn.go:
```diff
@@ -229,7 +229,7 @@ func (pc *PacketConn) SetReadDeadline(t time.Time) error {
 
 // SetWriteDeadline sets the write deadline.
-func (pc *PacketConn) SetWriteDeadline(t time.Time) error {
+func (pc *PacketConn) SetWriteDeadline(_ time.Time) error {
```
interface

AaronH88 commented on pkg/workceptor/kubernetes.go:
```diff
@@ -454,7 +452,7 @@ func (kw *kubeUnit) runWorkUsingLogger() {
 			var err error
 			for retries := 5; retries > 0; retries-- {
-				err = exec.Stream(remotecommand.StreamOptions{
+				err = exec.StreamWithContext(context.Background(), remotecommand.StreamOptions{
```
Nice, `kw` has the correct context, please use `kw.ctx` :smile:

AaronH88 commented on pkg/workceptor/remote_work.go:
```diff
@@ -219,7 +219,7 @@ func (rw *remoteUnit) startRemoteUnit(ctx context.Context, conn net.Conn, reader
 // cancelOrReleaseRemoteUnit makes a single attempt to cancel or release a remote unit.
 func (rw *remoteUnit) cancelOrReleaseRemoteUnit(ctx context.Context, conn net.Conn, reader *bufio.Reader,
-	release bool, force bool,
+	release bool, _ bool,
```
Nice, this can be removed :smile: 

Also the calls to it can remove the final bool please

AaronH88 commented on pkg/workceptor/workunitbase.go:
```diff
@@ -117,7 +117,7 @@ func (bwu *BaseWorkUnit) Debug(format string, v ...interface{}) {
 
 // SetFromParams sets the in-memory state from parameters.
-func (bwu *BaseWorkUnit) SetFromParams(params map[string]string) error {
+func (bwu *BaseWorkUnit) SetFromParams(_ map[string]string) error {
```
interface, this is correct

AaronH88 commented on tests/functional/mesh/lib.go:
```diff
@@ -770,12 +753,11 @@ type NativeBackend interface {
 }
 
-type backendInfo struct {
+type BackendInfo struct {
```
This seems fine :smile:

AaronH88 commented on tests/functional/mesh/mesh_test.go:
```diff
@@ -182,8 +182,9 @@ func TestTraceroute(t *testing.T) {
 
 // Test that a mesh starts and that connections are what we expect.
+//
+//nolint:tparallel
```
This is still true, nice work

AaronH88 commented:
Great work here! Some small tweaks and we should be ready :smile:

ToxicGLaDOS committed:
Fix linter fixup mistakes

Use correct context for `exec.StreamWithContext`, remove final bool
in cancelOrReleaseRemoteUnit altogether and fix newline spacing.

ToxicGLaDOS commented:
@AaronH88 Okay I think that covers the issues. Let me know if there's more to be done!

ToxicGLaDOS commented:
Huh... I see a linter error has appeared. I have the linter set to a pre-commit hook and I ran it manually before I committed. Maybe I'm on a different version of the linter? Anyways, I'll fix that up and squash it into the previous commit.

ToxicGLaDOS commented:
Oh but the CI doesn't give me a line number 🤔

AaronH88 commented:
> Oh but the CI doesn't give me a line number thinking

I think I can see more than you:

Check failure on line 328 in pkg/workceptor/kubernetes.go 
GitHub Actions
/ lint-receptor

return with no blank line before (nlreturn)

ToxicGLaDOS commented:
> > Oh but the CI doesn't give me a line number thinking
> 
> I think I can see more than you:
> 
> Check failure on line 328 in pkg/workceptor/kubernetes.go GitHub Actions / lint-receptor
> 
> return with no blank line before (nlreturn)

Huh... I'm not sure what to make of that. There isn't a return on that line? That seems to be the end of the `kw.Warning(...)` call.

AaronH88 commented:
> > > Oh but the CI doesn't give me a line number thinking
> > 
> > 
> > I think I can see more than you:
> > Check failure on line 328 in pkg/workceptor/kubernetes.go GitHub Actions / lint-receptor
> > return with no blank line before (nlreturn)
> 
> Huh... I'm not sure what to make of that. There isn't a return on that line? That seems to be the end of the `kw.Warning(...)` call.

That is very strange! 
Can you add a line break to see if it resolves please?

If it dosent, it might be a false positive and we should ignore.

ToxicGLaDOS commented:
I added a newline before the kw.Warning call. If that doesn't work I'd like to try rebasing onto devel. Maybe it has something to do with an interaction between my commit and something that's on devel that I don't have on my local machine yet 🤷.

ToxicGLaDOS commented:
Ah it was the rebase after all! I think this one will work 😅

AaronH88 commented:
> Ah it was the rebase after all! I think this one will work sweat_smile

Excellent! 

Ill give it one final review in the morning and merge, thanks so much for this!

AaronH88 reviewed (approved):
lgtm

AaronH88 commented:
Sorry for the delay in this, I needed to get more permissions to change the required tests for our workflows.

I also spoke with our Ansible stakeholders to ensure that this wont cause any problems and I have been asked to make a release now, before this PR gets merged.
So, I should be in a position to merge this in the next few days once this is all done :smile:

ToxicGLaDOS committed:
Fix merge conflict
